auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```diff
@@ -671,7 +708,166 @@ private MethodSpec createReadMethod(ClassName className,
+      ClassName jsonAdapter,
+      List<TypeVariableName> typeParams,
+      BuilderContext builderContext,
+      ProcessingEnvironment processingEnvironment) {
+    ParameterSpec jsonReader = ParameterSpec.builder(JsonReader.class, "jsonReader").build();
+    MethodSpec.Builder readMethod = MethodSpec.methodBuilder("read")
+        .addAnnotation(Override.class)
+        .addAnnotation(AnnotationSpec.builder(SuppressWarnings.class)
+            .addMember("value", "\"unchecked\"")
+            .build())
+        .addModifiers(PUBLIC)
+        .returns(autoValueClassName)
+        .addParameter(jsonReader)
+        .addException(IOException.class);
+
+    ClassName token = ClassName.get(JsonToken.NULL.getDeclaringClass());
+
+    readMethod.beginControlFlow("if ($N.peek() == $T.NULL)", jsonReader, token);
+    readMethod.addStatement("$N.nextNull()", jsonReader);
+    readMethod.addStatement("return null");
+    readMethod.endControlFlow();
+
+
+    // Instantiate the AutoValue'd class builder
+    FieldSpec builder = FieldSpec.builder(ClassName.get(builderContext.builderType()), "builder")
+        .build();
+
+    Set<ExecutableElement> builderMethods = builderContext.builderMethods();
+
+    if (builderMethods.size() == 0) {
+      // If no builder method defined, instantiate directly.
+      readMethod.addStatement("$T $N = new $T.$L()", builder.type, builder, className,
+          builderContext.builderType().getSimpleName());
+    } else {
+      ExecutableElement builderMethod;
+      if (builderMethods.size() == 1) {
+        // If there is only 1, use it.
+        builderMethod = builderMethods.stream().findFirst().get();
+      } else {
+        // Otherwise, find the only builder method that is annotated.
+        Set<ExecutableElement> annotatedMethods = builderMethods.stream()
+            .filter(e -> MoreElements.isAnnotationPresent(e, AutoValueGsonBuilder.class))
+            .collect(Collectors.toSet());
+
+        if (annotatedMethods.size() == 0) {
+          processingEnvironment.getMessager()
+              .printMessage(
+                  Diagnostic.Kind.ERROR,
+                  "Too many builder methods in " + className.simpleName() + ". Annotate the "
+                      + "builder this TypeAdapter should use with @AutoValueGsonBuilder."
+              );
+          throw new IllegalStateException();
+        } else if (annotatedMethods.size() == 1) {
+          builderMethod = annotatedMethods.stream().findFirst().get();
+        } else {
+          processingEnvironment.getMessager()
+              .printMessage(
+                  Diagnostic.Kind.ERROR,
+                  "Too many @AutoValueGsonBuilder annotations",
+                  annotatedMethods.stream().findFirst().get()
+              );
+          throw new IllegalStateException();
+        }
+      }
+
+      readMethod.addStatement("$T $N = $T.$L", builder.type, builder, autoValueClassName, builderMethod);
```

Comment:
Does JavaPoet automatically turn an ExecutableElement into its name + `()` here for `%L`?

Reply:
It does. I think I could also do the following. Not sure which is better practice.
```suggestion
      readMethod.addStatement("$T $N = $T.$N()", builder.type, builder, autoValueClassName, builderMethod.name);
```

Reply:
We can leave it as-is :+1: